pkg/proxy/dns.go

```diff
@@ -0,0 +1,455 @@
+const headlessSrvPortPathFormat = "%s." + srvPortPathFormat
+
+type serviceRecord interface {
+	getAllEntries() ([]skymsg.Service, bool)
+	getEntriesFor(segments []string) ([]skymsg.Service, bool)
+	confirm()
+}
+
+type normalServiceRecord struct {
+	portRecords  map[string]skymsg.Service
+	normalRecord skymsg.Service
+}
+
+func (rec *normalServiceRecord) getAllEntries() ([]skymsg.Service, bool) {
+	entries := make([]skymsg.Service, 0, len(rec.portRecords)+1)
+	for _, entry := range rec.portRecords {
+		entries = append(entries, entry)
+	}
+	return append(entries, rec.normalRecord), true
+}
+
+func (rec *normalServiceRecord) getEntriesFor(segments []string) ([]skymsg.Service, bool) {
+	switch len(segments) {
+	case 1:
+		// ns
+		fallthrough
+	case 2:
+		// name.ns
+		return rec.getAllEntries()
+	case 4:
+		// _port._proto.name.ns
+		entry, ok := rec.portRecords[segments[0]+"."+segments[1]]
+		return []skymsg.Service{entry}, ok
+	default:
+		// we don't support > 4 or _proto.name.ns
+		return nil, false
+	}
+}
+
+func (rec *normalServiceRecord) confirm() {}
+
+type headlessServiceRecord struct {
+	endpointRecords map[string]skymsg.Service
+	portRecords     map[string][]skymsg.Service
+	confirmed       bool
+}
+
+func (rec *headlessServiceRecord) getAllEntries() ([]skymsg.Service, bool) {
+	entries := make([]skymsg.Service, 0, len(rec.portRecords)+len(rec.portRecords)*len(rec.endpointRecords))
+	for _, entry := range rec.endpointRecords {
+		entries = append(entries, entry)
+	}
+	for _, entry := range rec.portRecords {
+		entries = append(entries, entry...)
+	}
+	return entries, true
+}
+
+func (rec *headlessServiceRecord) getEntriesFor(segments []string) ([]skymsg.Service, bool) {
+	switch len(segments) {
+	case 1:
+		// ns
+		fallthrough
+	case 2:
+		// name.ns
+		return rec.getAllEntries()
+	case 3:
+		if strings.HasPrefix(segments[0], "_") {
+			// _proto.name.ns
+			return nil, false
+		}
+		// ep.name.ns
+		entry, ok := rec.endpointRecords[segments[0]]
+		return []skymsg.Service{entry}, ok
+	case 4:
+		// _port._proto.name.ns
+		entries, ok := rec.portRecords[segments[0]+"."+segments[1]]
+		return entries, ok
+	default:
+		// we don't support > 5 or _proto.name.ns
+		return nil, false
+	}
+}
+
+func (rec *headlessServiceRecord) confirm() {
+	rec.confirmed = true
+}
+
+type DNSHandler struct {
+	mu             sync.RWMutex
+	serviceRecords map[types.NamespacedName]serviceRecord
+	isHeadless     map[types.NamespacedName]bool
+}
+
+func NewDNSHandler() *DNSHandler {
+	return &DNSHandler{
+		serviceRecords: make(map[types.NamespacedName]serviceRecord),
+		isHeadless:     make(map[types.NamespacedName]bool),
+	}
+}
+
+type DNSServiceHandler struct {
+	*DNSHandler
+}
+
+type DNSEndpointHandler struct {
+	*DNSHandler
+}
+
+func getHash(text string) string {
+	h := fnv.New32a()
+	h.Write([]byte(text))
+	return fmt.Sprintf("%x", h.Sum32())
+}
+
+func (handler *DNSServiceHandler) OnUpdate(services []api.Service) {
```

Comment:
The DNS service resolver in Openshift could in theory be used whole cloth.
In the long run, the proxy should be using cache.Store.  It may be good to
split that refactor if it looks feasible (proxy to cache.Store first, then
DNS).  Im pretty sure the refactor is possible, but there may be a wrinkle
I'm missing.  You will need the reverse cluster IP indexer anyway.

On Jul 21, 2015, at 10:31 AM, Solly <notifications@github.com> wrote:

In pkg/proxy/dns.go
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11599#discussion_r35108136>
:

> +
> +type DNSServiceHandler struct {
> +	*DNSHandler
> +}
> +
> +type DNSEndpointHandler struct {
> +	*DNSHandler
> +}
> +
> +func getHash(text string) string {
> +	h := fnv.New32a()
> +	h.Write([]byte(text))
> +	return fmt.Sprintf("%x", h.Sum32())
> +}
> +
> +func (handler *DNSServiceHandler) OnUpdate(services []api.Service) {

Ah, ok. The original suggestion was that "proxy already has the
information", so I tried to mirror the existing code in proxier.go and
endpoints.go. I will take a look at cache.Store and friends.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11599/files#r35108136>
.